src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1319,6 +1321,150 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
   return r;
 }
 
+int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &current,
+                                 const boost::optional<Snapshot>& prev) {
+  if (!prev) {
+    derr << "Invalid previous snapshot" << dendl;
+    return -ENODATA;
+  }
+
+  dout(20) << ": incremental sync check from prev=" << prev << dendl;
+
+  FHandles fh;
+  int r = pre_sync_check_and_open_handles(dir_root, current, prev, &fh);
+  if (r < 0) {
+    dout(5) << ": cannot proceeed with sync: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  BOOST_SCOPE_EXIT_ALL( (this)(&fh) ) {
+    post_sync_close_handles(fh);
+  };
+
+  // record that we are going to "dirty" the data under this directory root
+  auto snap_id_str{stringify(current.second)};
+  r = ceph_setxattr(m_remote_mount, dir_root.c_str(), "ceph.mirror.dirty_snap_id",
+                    snap_id_str.c_str(), snap_id_str.size(), 0);
+  if (r < 0) {
+    derr << ": error setting \"ceph.mirror.dirty_snap_id\" on dir_root=" << dir_root
+         << ": " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  ceph_snapdiff_info sd_info;
+  ceph_snapdiff_entry_t sd_entry;
+  std::string epath = ".";
+  std::string npath = "", nabs_path = "", nname = "";
+  int ntype = 0;
+
+  //the queue of SyncDiffEntry items (directories) to be synchronized
```

Comment:
I kind of recall mentioning this a while back, but what's the intention of accumulating the entries in a queue?

Reply:
It's changed to queue because the snapdiff code needs the current entry (parent) to pop out once it's processed. When we try to call the pop(), the stack's been already emplaced with the child entries for later processing. So a pop() call would be removing the top entry, which would be a child entry mostly. But we wanted to remove the current entry which is at the bottom of the stack. So if we use stack in snapdiff iteration, what we wanted to remove is at the bottom of the stack. This issue has been resolved by introducing a queue.

Reply:
Why can't we do something like:

```
while (no more items to proces) {
          while (true) {
              r = readdir_snap_diff(...)
              // handle error

              if (dir) {
                 // put it on the top of stack
             } else {
                 // non dir - sync!
             }
           }
}
```

In short, even when using the snapdiff API, using a (sync) stack would suffice.

Reply:
This^ is what I'm doing. Please check out the code again.

> In short, even when using the snapdiff API, using a (sync) stack would suffice.

There's nothing wrong with using queue too. The negative side is already explained in the above comment. If there's a way to remove the bottom element of stack that would help.

Reply:
> This^ is what I'm doing. Please check out the code again.
> 
> > In short, even when using the snapdiff API, using a (sync) stack would suffice.
> 
> There's nothing wrong with using queue too. The negative side is already explained in the above comment. If there's a way to remove the bottom element of stack that would help. Otherwise, queue is required here.

There isn't anything wrong - that wasn't my point. If you see how do_synchronize() is implemented without this change, using a stack worked just fine and the same can be done when using the snapdiff API. Could you explain what difference in approach lies with using the snapdiff as compared to iterating the entire directory?

Reply:
The existing `do_synchronize()` also needs queue, in my opinion. But as long as it works and tested using stack since long time, let it be there. It avoids the issue somehow.

In the new `do_synchronize()`,
My requirement here is to pop elements from the bottom of the container (FIFO) and that works perfectly using queue. I'm writing this based on my debugging results of each snapdiff iteration. When I used stack earlier, the top element only gets popped out from the stack and then proceeds to an unnecessary syncing (processing directories again that are already processed), as the opened directory (say dot `.`) is already there at the bottom of the stack. The newly found sub directories added to the top. By using queue the opened directory gets pop out at the end of the iteration (expected), no need to do anything special for this. It does the same for each directory syncing. Hope that clarifies.

Reply:
> The existing `do_synchronize()` also needs queue, in my opinion. But as long as it works and tested using stack since long time, let it be there. It avoids the issue somehow.

The issues is not avoided "somehow", but rather the directory tree traversal is implemented with a stack in mind.

> 
> In the new `do_synchronize()`, My requirement here is to pop elements from the bottom of the container (FIFO) and that works perfectly using queue. I'm writing this based on my debugging results of each snapdiff iteration. When I used stack earlier, the top element only gets popped out from the stack and then proceeds to an unnecessary syncing (processing directories again that are already processed), as the opened directory (say dot `.`) is already there at the bottom of the stack.

You should be filtering out `.` and `..` when iterating a directory (diff), isn't it?

> The newly found sub directories added to the top. By using queue the opened directory gets pop out at the end of the iteration (expected), no need to do anything special for this. It does the same for each directory syncing. Hope that clarifies.

Are you calling pushing all the diff items into the queue all at once?

Reply:
Got it. Updated the code using stack.